akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala

```diff
@@ -345,10 +348,32 @@ object StreamLayout {
       if (s != shape) throw new UnsupportedOperationException("cannot replace the shape of the EmptyModule")
       else this
 
-    override def compose(that: Module): Module = that
-
-    override def compose[A, B, C](that: Module, f: (A, B) ⇒ C): Module =
-      throw new UnsupportedOperationException("It is invalid to combine materialized value with EmptyModule")
+    override def compose(that: Module): Module = compose(that, scaladsl.Keep.left)
+
+    override def compose[A, B, C](that: Module, f: (A, B) ⇒ C): Module = {
+      if (f == scaladsl.Keep.right) {
+        that
+      } else if (f == scaladsl.Keep.left) {
```

Comment:
Also, the `compose` method on Empty module is indirectly called by the GraphDSL, no user provided function can and up here (it throws on anything that is not a Keep.left or right, used to be throw on any usage of this method, see diff).